File: apps/ios/MiWarpMobile/MiWarpMobile/Core/MiWarpEventReducer.swift
import Foundation

// MARK: - Display Message

struct DisplayMessage: Identifiable {
    let id: String
    var role: MessageRole
    var content: String
    var timestamp: Date
    var isStreaming: Bool
    var toolCalls: [DisplayToolCall]
    var thinking: String?

    enum MessageRole {
        case user
        case assistant
        case system
    }
}

// MARK: - Display Tool Call

struct DisplayToolCall: Identifiable {
    let id: String
    let toolName: String
    var inputPreview: String?
    var output: String?
    var isComplete: Bool
    var isError: Bool
    var isExpanded: Bool
}

// MARK: - Permission Request

struct PendingPermission: Identifiable {
    let id: String // requestId
    let runId: String
    let toolName: String
    let toolUseId: String
    let toolInput: AnyCodable?
    let description: String?
}

// MARK: - Usage Summary

struct UsageSummary {
    var inputTokens: Int = 0
    var outputTokens: Int = 0
    var cacheReadTokens: Int = 0
    var cacheWriteTokens: Int = 0
    var costUsd: Double = 0
}

// MARK: - Protocol Recovery State

struct ProtocolRecoveryState: Equatable {
    var notice: String?
    var isRecovering: Bool = false
    var showReloadAction: Bool = false
}

// MARK: - Event Reducer

@Observable
@MainActor
final class MiWarpEventReducer {
    private(set) var messages: [DisplayMessage] = []
    private(set) var pendingPermissions: [PendingPermission] = []
    private(set) var usage: UsageSummary = UsageSummary()
    private(set) var currentStatus: RunStatus = .idle
    private(set) var sessionAgent: String?
    private(set) var sessionModel: String?
    private(set) var lastSeq: Int = 0
    private(set) var streamingMessageId: String?
    private(set) var protocolRecovery = ProtocolRecoveryState()
    /// Tracks all in-flight streaming message IDs so handleMessageComplete can find them.
    private var pendingStreamingIds: Set<String> = []

    private struct ToolLocation {
        let messageIndex: Int
        let toolIndex: Int
    }

    /// Steady-state O(1) lookup for tool delta/end events in long conversations.
    private var toolLocations: [String: ToolLocation] = [:]
    private var seenSeqs: Set<Int> = []

    // MARK: - Process Event

    func processEvent(_ event: BusEvent) {
        // Dedup
        guard !seenSeqs.contains(event.seq) else { return }
        seenSeqs.insert(event.seq)
        lastSeq = max(lastSeq, event.seq)

        // Cap seenSeqs to prevent unbounded growth
        if seenSeqs.count > 8192 {
            seenSeqs = Set(seenSeqs.filter { $0 > lastSeq - 8192 })
        }

        switch event.payload {
        case .userMessage(let payload):
            handleUserMessage(payload)

        case .messageDelta(let payload):
            handleMessageDelta(payload)

        case .messageComplete(let payload):
            handleMessageComplete(payload)

        case .toolStart(let payload):
            handleToolStart(payload)

        case .toolEnd(let payload):
            handleToolEnd(payload)

        case .runState(let payload):
            handleRunState(payload)

        case .usageUpdate(let payload):
            handleUsageUpdate(payload)

        case .thinkingDelta(let payload):
            handleThinkingDelta(payload)

        case .permissionPrompt(let payload):
            handlePermissionPrompt(payload, runId: event.runId)

        case .permissionDenied(let payload):
            handlePermissionDenied(payload)

        case .systemStatus(let payload):
            handleSystemStatus(payload)

        case .toolInputDelta(let payload):
            handleToolInputDelta(payload)

        case .commandOutput(let payload):
            handleCommandOutput(payload)

        case .sessionInit(let payload):
            handleSessionInit(payload)

        case .sessionRecovering(let payload):
            handleSessionRecovering(payload)

        case .sessionRecovered(let payload):
            handleSessionRecovered(payload)

        case .protocolDesync(let payload):
            handleProtocolDesync(payload)

        case .fullReload:
            // Server requests full state reset — clear and re-fetch
            reset()

        default:
            break
        }
    }

    // MARK: - Load History

    func loadHistory(_ events: [BusEvent]) {
        reset()
        for event in events.sorted(by: { $0.seq < $1.seq }) {
            processEvent(event)
        }
    }

    func reset() {
        messages.removeAll()
        pendingPermissions.removeAll()
        usage = UsageSummary()
        currentStatus = .idle
        sessionAgent = nil
        sessionModel = nil
        lastSeq = 0
        seenSeqs.removeAll()
        streamingMessageId = nil
        pendingStreamingIds.removeAll()
        toolLocations.removeAll()
        protocolRecovery = ProtocolRecoveryState()
    }

    // MARK: - Handlers

    private func handleSessionInit(_ payload: SessionInitPayload) {
        sessionAgent = payload.agent
        sessionModel = payload.model
    }

    private func handleSessionRecovering(_ payload: SessionRecoveringPayload) {
        let seconds = max(1, Int((payload.deadlineMs ?? 5000) / 1000))
        protocolRecovery = ProtocolRecoveryState(
            notice: String(format: String(localized: "protocol.sessionRecovering"), seconds),
            isRecovering: true,
            showReloadAction: false
        )
    }

    private func handleSessionRecovered(_ payload: SessionRecoveredPayload) {
        if payload.ok == true {
            protocolRecovery = ProtocolRecoveryState()
            return
        }

        protocolRecovery = ProtocolRecoveryState(
            notice: String(localized: "protocol.sessionRecoveryFailed"),
            isRecovering: false,
            showReloadAction: true
        )
        currentStatus = .failed
    }

    private func handleProtocolDesync(_ payload: ProtocolDesyncPayload) {
        _ = payload
        protocolRecovery = ProtocolRecoveryState(
            notice: String(localized: "protocol.desyncToast"),
            isRecovering: false,
            showReloadAction: true
        )
        currentStatus = .failed
    }

    func clearProtocolRecoveryNotice() {
        protocolRecovery.notice = nil
        protocolRecovery.showReloadAction = false
    }

    private func handleUserMessage(_ payload: UserMessagePayload) {
        guard let content = payload.content else { return }
        let msg = DisplayMessage(
            id: "user-\(lastSeq)",
            role: .user,
            content: content,
            timestamp: Date(),
            isStreaming: false,
            toolCalls: []
        )
        messages.append(msg)
    }

    private func handleMessageDelta(_ payload: MessageDeltaPayload) {
        // Use provided messageId, or generate a UUID — never a static string that collides across messages.
        let msgId = payload.messageId ?? UUID().uuidString
        let delta = payload.delta ?? payload.content ?? ""

        if streamingMessageId == msgId {
            // Append to existing streaming message
            if let index = messages.lastIndex(where: { $0.id == msgId }) {
                messages[index].content += delta
            }
        } else {
            // Start new streaming message
            streamingMessageId = msgId
            pendingStreamingIds.insert(msgId)
            let msg = DisplayMessage(
                id: msgId,
                role: .assistant,
                content: delta,
                timestamp: Date(),
                isStreaming: true,
                toolCalls: []
            )
            messages.append(msg)
        }

    }

    private func handleMessageComplete(_ payload: MessageCompletePayload) {
        let msgId: String?
        if let id = payload.messageId {
            msgId = id
        } else if let current = streamingMessageId, pendingStreamingIds.contains(current) {
            msgId = current
        } else {
            msgId = nil
        }
        guard let id = msgId else { return }
        if let index = messages.lastIndex(where: { $0.id == id }) {
            messages[index].isStreaming = false
            if let content = payload.content {
                messages[index].content = content
            }
        }
        pendingStreamingIds.remove(id)
        streamingMessageId = pendingStreamingIds.first
    }

    private func handleToolStart(_ payload: ToolStartPayload) {
        guard let toolName = payload.toolName else { return }
        let toolId = payload.toolUseId ?? UUID().uuidString

        // Skip if this toolUseId is already present (prevents duplicate IDs from retried events).
        if findToolLocation(toolId) != nil { return }

        var inputPreview: String?
        if let input = payload.input {
            if let dict = input.dictValue {
                if let data = try? JSONSerialization.data(withJSONObject: dict, options: .fragmentsAllowed),
                   let str = String(data: data, encoding: .utf8) {
                    inputPreview = String(str.prefix(200))
                }
            } else if let str = input.stringValue {
                inputPreview = String(str.prefix(200))
            }
        }

        let toolCall = DisplayToolCall(
            id: toolId,
            toolName: toolName,
            inputPreview: inputPreview,
            output: nil,
            isComplete: false,
            isError: false,
            isExpanded: false
        )

        // Attach to last assistant message or create new one
        if let lastIndex = messages.indices.last, messages[lastIndex].role == .assistant {
            messages[lastIndex].toolCalls.append(toolCall)
            toolLocations[toolId] = ToolLocation(
                messageIndex: lastIndex,
                toolIndex: messages[lastIndex].toolCalls.count - 1
            )
        } else {
            let msg = DisplayMessage(
                id: "tool-\(toolId)",
                role: .assistant,
                content: "",
                timestamp: Date(),
                isStreaming: false,
                toolCalls: [toolCall]
            )
            messages.append(msg)
            toolLocations[toolId] = ToolLocation(
                messageIndex: messages.count - 1,
                toolIndex: 0
            )
        }
    }

    private func handleToolEnd(_ payload: ToolEndPayload) {
        guard let toolId = payload.toolUseId else { return }
        let outputString: String? = {
            guard let output = payload.output else { return nil }
            if let str = output.stringValue { return str }
            if let dict = output.dictValue,
               let data = try? JSONSerialization.data(withJSONObject: dict, options: [.fragmentsAllowed, .sortedKeys]),
               let str = String(data: data, encoding: .utf8) {
                return str
            }
            if let arr = output.arrayValue,
               let data = try? JSONSerialization.data(withJSONObject: arr, options: [.fragmentsAllowed, .sortedKeys]),
               let str = String(data: data, encoding: .utf8) {
                return str
            }
            return nil
        }()

        guard let location = findToolLocation(toolId) else { return }
        messages[location.messageIndex].toolCalls[location.toolIndex].output = outputString
        messages[location.messageIndex].toolCalls[location.toolIndex].isComplete = true
        messages[location.messageIndex].toolCalls[location.toolIndex].isError = payload.status != "success"
    }

    private func handleRunState(_ payload: RunStatePayload) {
        if let status = payload.status {
            currentStatus = status
        }
    }

    private func handleUsageUpdate(_ payload: UsageUpdatePayload) {
        usage.inputTokens += payload.inputTokens ?? 0
        usage.outputTokens += payload.outputTokens ?? 0
        usage.cacheReadTokens += payload.cacheReadTokens ?? 0
        usage.cacheWriteTokens += payload.cacheWriteTokens ?? 0
        if let cost = payload.costUsd {
            usage.costUsd += cost
        }
    }

    private func handleThinkingDelta(_ payload: ThinkingDeltaPayload) {
        guard let text = payload.text else { return }
        if let lastIndex = messages.indices.last, messages[lastIndex].role == .assistant {
            messages[lastIndex].thinking = (messages[lastIndex].thinking ?? "") + text
        }
    }

    private func handlePermissionPrompt(_ payload: PermissionPromptPayload, runId: String) {
        guard let requestId = payload.requestId, let toolName = payload.toolName else { return }
        let permission = PendingPermission(
            id: requestId,
            runId: runId,
            toolName: toolName,
            toolUseId: payload.toolUseId ?? "",
            toolInput: payload.toolInput,
            description: payload.description
        )
        pendingPermissions.append(permission)
        currentStatus = .waitingApproval
    }

    private func handlePermissionDenied(_ payload: PermissionDeniedPayload) {
        if let toolUseId = payload.toolUseId, !toolUseId.isEmpty {
            pendingPermissions.removeAll { $0.toolUseId == toolUseId }
        } else if let toolName = payload.toolName {
            pendingPermissions.removeAll { $0.toolName == toolName }
        }
        if pendingPermissions.isEmpty { currentStatus = .running }
    }

    private func handleSystemStatus(_ payload: SystemStatusPayload) {
        guard let status = payload.status, !status.isEmpty else { return }
        let msg = DisplayMessage(
            id: "system-\(lastSeq)",
            role: .system,
            content: status,
            timestamp: Date(),
            isStreaming: false,
            toolCalls: []
        )
        messages.append(msg)
    }

    private func handleToolInputDelta(_ payload: ToolInputDeltaPayload) {
        guard
            let toolId = payload.toolUseId,
            let delta = payload.partialJson,
            let location = findToolLocation(toolId)
        else { return }

        messages[location.messageIndex].toolCalls[location.toolIndex].inputPreview =
            (messages[location.messageIndex].toolCalls[location.toolIndex].inputPreview ?? "") + delta
    }

    private func findToolLocation(_ toolId: String) -> ToolLocation? {
        if let cached = toolLocations[toolId],
           messages.indices.contains(cached.messageIndex),
           messages[cached.messageIndex].toolCalls.indices.contains(cached.toolIndex),
           messages[cached.messageIndex].toolCalls[cached.toolIndex].id == toolId {
            return cached
        }

        // Defensive fallback keeps behavior correct if future code mutates message ordering.
        for messageIndex in messages.indices {
            if let toolIndex = messages[messageIndex].toolCalls.firstIndex(where: { $0.id == toolId }) {
                let location = ToolLocation(messageIndex: messageIndex, toolIndex: toolIndex)
                toolLocations[toolId] = location
                return location
            }
        }
        toolLocations.removeValue(forKey: toolId)
        return nil
    }

    private func handleCommandOutput(_ payload: CommandOutputPayload) {
        guard let content = payload.content, !content.isEmpty else { return }
        let msg = DisplayMessage(
            id: "cmdout-\(lastSeq)",
            role: .system,
            content: content,
            timestamp: Date(),
            isStreaming: false,
            toolCalls: []
        )
        messages.append(msg)
    }

    // MARK: - Remove Permission

    func removePermission(requestId: String) {
        pendingPermissions.removeAll { $0.id == requestId }
    }
}

File: apps/android/MiWarpMobile/app/src/main/java/com/miwarp/mobile/reducer/MiWarpEventReducer.kt
package com.miwarp.mobile.reducer

import com.miwarp.mobile.model.BusEvent
import com.miwarp.mobile.model.ChatMessage
import com.miwarp.mobile.model.MessageRole
import com.miwarp.mobile.model.PermissionRequest
import com.miwarp.mobile.model.RunStatus
import com.miwarp.mobile.model.ToolCallInfo
import com.miwarp.mobile.model.UsageSummary
import java.util.UUID
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.contentOrNull

/**
 * Reduces a stream of BusEvents into ChatMessage list + derived state.
 * Mirrors the iOS event reducer logic.
 */
class MiWarpEventReducer {

    private val messages = mutableListOf<ChatMessage>()
    private val seenSeqs = mutableSetOf<Long>()
    private val pendingPermissions = mutableListOf<PermissionRequest>()
    private var usage = UsageSummary()
    private var currentStatus: RunStatus = RunStatus.Idle
    private var lastSeq: Long = 0L
    private var streamingMessageId: String? = null
    private val messageAccumulator = mutableMapOf<String, String>() // messageId -> accumulated text

    fun getMessages(): List<ChatMessage> = messages.toList()

    fun getPendingPermissions(): List<PermissionRequest> = pendingPermissions.toList()

    fun getUsage(): UsageSummary = usage.copy()

    fun getCurrentStatus(): RunStatus = currentStatus

    fun getLastSeq(): Long = lastSeq

    fun reduce(event: BusEvent): ReductionResult {
        if (event.seq > 0 && !seenSeqs.add(event.seq)) {
            return ReductionResult(
                messages = getMessages(),
                changed = false,
                pendingPermissions = getPendingPermissions(),
                usage = getUsage(),
                currentStatus = currentStatus,
            )
        }

        if (event.seq > lastSeq) lastSeq = event.seq

        // Cap seenSeqs to prevent unbounded growth (keep last 8192 entries)
        if (seenSeqs.size > 8192) {
            seenSeqs.retainAll { it > lastSeq - 8192 }
        }

        var changed = false

        when (event) {
            is BusEvent.MessageDelta -> {
                changed = handleMessageDelta(event)
            }
            is BusEvent.MessageComplete -> {
                changed = handleMessageComplete(event)
            }
            is BusEvent.UserMessage -> {
                changed = handleUserMessage(event)
            }
            is BusEvent.ToolStart -> {
                changed = handleToolStart(event)
            }
            is BusEvent.ToolEnd -> {
                changed = handleToolEnd(event)
            }
            is BusEvent.ToolInputDelta -> {
                changed = handleToolInputDelta(event)
            }
            is BusEvent.ToolProgress -> {
                changed = handleToolProgress(event)
            }
            is BusEvent.ThinkingDelta -> {
                changed = handleThinkingDelta(event)
            }
            is BusEvent.PermissionPrompt -> {
                pendingPermissions.add(
                    PermissionRequest(
                        requestId = event.requestId,
                        runId = event.runId,
                        toolName = event.toolName,
                        toolUseId = event.toolUseId,
                        description = event.description,
                        options = event.options,
                    )
                )
                currentStatus = RunStatus.WaitingApproval
                changed = true
            }
            is BusEvent.PermissionDenied -> {
                if (event.toolUseId.isNotEmpty()) {
                    pendingPermissions.removeAll { it.toolUseId == event.toolUseId }
                } else {
                    pendingPermissions.removeAll { it.toolName == event.toolName }
                }
                if (pendingPermissions.isEmpty()) currentStatus = RunStatus.Running
                changed = true
            }
            is BusEvent.RunState -> {
                currentStatus = event.status
                changed = true
            }
            is BusEvent.UsageUpdate -> {
                usage.inputTokens += event.inputTokens
                usage.outputTokens += event.outputTokens
                usage.cacheReadTokens += event.cacheReadTokens
                usage.cacheWriteTokens += event.cacheWriteTokens
                usage.costUsd += event.costUsd
                changed = true
            }
            is BusEvent.SystemStatus -> {
                val status = event.status
                if (!status.isNullOrEmpty()) {
                    messages.add(
                        ChatMessage(
                            id = "system-$lastSeq",
                            role = MessageRole.System,
                            text = status,
                        )
                    )
                    changed = true
                }
            }
            is BusEvent.FullReload -> {
                clear()
                changed = true
            }
            is BusEvent.CompactBoundary -> {
                // Mark the boundary; content continues after
            }
            is BusEvent.CommandOutput -> {
                if (!event.content.isNullOrEmpty()) {
                    messages.add(
                        ChatMessage(
                            id = "cmdout-$lastSeq",
                            role = MessageRole.System,
                            text = event.content,
                        )
                    )
                    changed = true
                }
            }
            else -> { /* Other events don't directly affect chat messages */ }
        }

        return ReductionResult(
            messages = getMessages(),
            changed = changed,
            pendingPermissions = getPendingPermissions(),
            usage = getUsage(),
            currentStatus = currentStatus,
        )
    }

    fun clear() {
        messages.clear()
        seenSeqs.clear()
        pendingPermissions.clear()
        usage = UsageSummary()
        currentStatus = RunStatus.Idle
        lastSeq = 0
        streamingMessageId = null
        messageAccumulator.clear()
    }

    fun removePermission(requestId: String) {
        pendingPermissions.removeAll { it.requestId == requestId }
    }

    private fun handleMessageDelta(event: BusEvent.MessageDelta): Boolean {
        val msgId = streamingMessageId ?: UUID.randomUUID().toString()
        val existing = messages.lastOrNull { it.role == MessageRole.Assistant && it.isStreaming }
        if (existing != null) {
            val index = messages.indexOf(existing)
            messages[index] = existing.copy(text = existing.text + event.text)
        } else {
            streamingMessageId = msgId
            messages.add(
                ChatMessage(
                    id = msgId,
                    role = MessageRole.Assistant,
                    text = event.text,
                    isStreaming = true,
                )
            )
        }
        messageAccumulator[msgId, default = ""] += event.text
        return true
    }

    private fun handleMessageComplete(event: BusEvent.MessageComplete): Boolean {
        val streaming = messages.lastOrNull { it.role == MessageRole.Assistant && it.isStreaming }
        if (streaming != null) {
            val index = messages.indexOf(streaming)
            val finalText = event.text ?: streaming.text
            messages[index] = streaming.copy(text = finalText, isStreaming = false)
            streamingMessageId = null
            messageAccumulator.remove(streaming.id)
            return true
        }
        return false
    }

    private fun handleUserMessage(event: BusEvent.UserMessage): Boolean {
        val text = try {
            event.payload?.jsonObject?.get("text")?.jsonPrimitive?.contentOrNull ?: ""
        } catch (_: Exception) { "" }
        messages.add(
            ChatMessage(
                id = "user-$lastSeq",
                role = MessageRole.User,
                text = text,
            )
        )
        return true
    }

    private fun handleToolStart(event: BusEvent.ToolStart): Boolean {
        val inputPreview = event.input?.toString()?.take(200) ?: ""
        val tool = ToolCallInfo(
            toolId = event.toolId,
            toolName = event.toolName,
            inputPreview = inputPreview,
            isRunning = true,
        )

        // Attach to the last assistant message
        val lastAssistant = messages.lastOrNull { it.role == MessageRole.Assistant }
        if (lastAssistant != null) {
            val index = messages.indexOf(lastAssistant)
            messages[index] = lastAssistant.copy(
                toolCalls = lastAssistant.toolCalls + tool,
            )
        }
        return true
    }

    private fun handleToolEnd(event: BusEvent.ToolEnd): Boolean {
        val outputStr = event.output?.toString()?.take(2000) ?: ""
        for (i in messages.indices.reversed()) {
            val msg = messages[i]
            if (msg.role == MessageRole.Assistant) {
                val toolIndex = msg.toolCalls.indexOfFirst { it.toolId == event.toolId }
                if (toolIndex >= 0) {
                    val updated = msg.toolCalls.toMutableList()
                    updated[toolIndex] = updated[toolIndex].copy(
                        output = outputStr,
                        isRunning = false,
                        isError = event.status != "success",
                        progress = -1.0,
                    )
                    messages[i] = msg.copy(toolCalls = updated)
                    return true
                }
            }
        }
        return false
    }

    private fun handleToolInputDelta(event: BusEvent.ToolInputDelta): Boolean {
        for (i in messages.indices.reversed()) {
            val msg = messages[i]
            if (msg.role == MessageRole.Assistant) {
                val toolIndex = msg.toolCalls.indexOfFirst { it.toolId == event.toolId }
                if (toolIndex >= 0) {
                    val updated = msg.toolCalls.toMutableList()
                    updated[toolIndex] = updated[toolIndex].copy(
                        inputPreview = updated[toolIndex].inputPreview + event.text,
                    )
                    messages[i] = msg.copy(toolCalls = updated)
                    return true
                }
            }
        }
        return false
    }

    private fun handleToolProgress(event: BusEvent.ToolProgress): Boolean {
        for (i in messages.indices.reversed()) {
            val msg = messages[i]
            if (msg.role == MessageRole.Assistant) {
                val toolIndex = msg.toolCalls.indexOfFirst { it.toolId == event.toolId }
                if (toolIndex >= 0) {
                    val updated = msg.toolCalls.toMutableList()
                    updated[toolIndex] = updated[toolIndex].copy(
                        progress = event.elapsedTimeSeconds,
                        progressMessage = "",
                    )
                    messages[i] = msg.copy(toolCalls = updated)
                    return true
                }
            }
        }
        return false
    }

    private fun handleThinkingDelta(event: BusEvent.ThinkingDelta): Boolean {
        val existing = messages.lastOrNull { it.role == MessageRole.Assistant && it.isStreaming }
        if (existing != null) {
            val index = messages.indexOf(existing)
            messages[index] = existing.copy(thinking = existing.thinking + event.text)
        } else {
            messages.add(
                ChatMessage(
                    id = UUID.randomUUID().toString(),
                    role = MessageRole.Assistant,
                    thinking = event.text,
                    isStreaming = true,
                )
            )
        }
        return true
    }
}

data class ReductionResult(
    val messages: List<ChatMessage>,
    val changed: Boolean,
    val pendingPermissions: List<PermissionRequest>,
    val usage: UsageSummary,
    val currentStatus: RunStatus,
)
